nighthawk/test/benchmark_http_client_test.cc

```diff
@@ -0,0 +1,289 @@
+
+    lorem_ipsum_config = filesystem.fileReadToEnd(Envoy::TestEnvironment::runfilesPath(
+        "nighthawk/test/test_data/benchmark_http_client_test_envoy.yaml"));
+    lorem_ipsum_config = Envoy::TestEnvironment::substitute(lorem_ipsum_config);
+  }
+
+  void SetUp() override {
+    ares_library_init(ARES_LIB_INIT_ALL);
+    Envoy::Event::Libevent::Global::initialize();
+    BaseIntegrationTest::initialize();
+  }
+
+  uint32_t getTestServerHostAndPort() { return lookupPort("listener_0"); }
+
+  uint32_t getTestServerHostAndSslPort() { return lookupPort("listener_1"); }
+
+  void TearDown() override {
+    client_->terminate();
+    test_server_.reset();
+    fake_upstreams_.clear();
+    tls_.shutdownGlobalThreading();
+    ares_library_cleanup();
+  }
+
+  void setupBenchmarkClient(const std::string uriPath, bool use_https, bool use_h2,
+                            uint32_t port = 0) {
+    if (port == 0) {
+      port = use_https ? getTestServerHostAndSslPort() : getTestServerHostAndPort();
+    }
+
+    client_ = std::make_unique<Client::BenchmarkClientHttpImpl>(
+        api_, *dispatcher_, std::make_unique<Envoy::Stats::IsolatedStoreImpl>(),
+        std::make_unique<StreamingStatistic>(), std::make_unique<StreamingStatistic>(),
+        fmt::format("{}://127.0.0.1:{}{}", use_https ? "https" : "http", port, uriPath), use_h2);
+  }
+
+  void testBasicFunctionality(const std::string uriPath, uint64_t max_pending,
+                              uint64_t connection_limit, bool use_https, bool use_h2,
+                              uint64_t amount_of_request) {
+    setupBenchmarkClient(uriPath, use_https, use_h2);
+
+    client_->set_connection_timeout(10s);
+    client_->set_max_pending_requests(max_pending);
+    client_->set_connection_limit(connection_limit);
+    EXPECT_TRUE(client_->initialize(runtime_));
+
+    uint64_t amount = amount_of_request;
+    uint64_t inflight_response_count = 0;
+
+    std::function<void()> f = [this, &inflight_response_count]() {
+      --inflight_response_count;
+      if (inflight_response_count == 0) {
+        dispatcher_->exit();
+      }
+    };
+
+    for (uint64_t i = 0; i < amount; i++) {
+      if (client_->tryStartOne(f)) {
+        inflight_response_count++;
+      }
+    }
+
+    EXPECT_EQ(max_pending, inflight_response_count);
+
+    dispatcher_->run(Envoy::Event::Dispatcher::RunType::Block);
+
+    EXPECT_EQ(0, client_->stream_reset_count());
+  }
+
+  std::string getNonZeroValuedCounters() {
+    Client::CounterFilter filter = [](std::string, uint64_t value) { return value > 0; };
+    return client_->countersToString(filter);
+  }
+
+  Envoy::Thread::ThreadFactoryImplPosix thread_factory_;
+  Envoy::Stats::IsolatedStoreImpl store_;
+  Envoy::Api::Impl api_;
+  Envoy::Event::DispatcherPtr dispatcher_;
+  Envoy::Runtime::RandomGeneratorImpl generator_;
+  ::testing::NiceMock<Envoy::ThreadLocal::MockInstance> tls_;
```

Comment:
Nit: probably worth a `using ::testing::NiceMock` at the top of file.

Reply:
Sure, I'd love to, I thought that `using` should be avoided according to the style guide, but I'm not sure where I picked that up. I propose we do that in a separate PR and to enforce consistency accross the board. Added to the track-for-future list.